### PR TITLE
refactor(`Message`): use raw data storage for messages

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Caches applicable message data within an internal `data` field. Makes props of the data getters in the class instead of storing them.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: packages/discord.js/src/structures/Message.js [hunk 1]
```diff
@@ -30,127 +30,401 @@ const Util = require('../util/Util');
...
+  }
+
+  /**
+   * Whether or not this message is pinned
+   * @type {?boolean}
+   */
+  get pinned() {
+    if (!this.data.pinned) return null;
+    return Boolean(this.data.pinned);
+  }
```
Comment (almeidx) [inline packages/discord.js/src/structures/Message.js:112]: This doesn't look right. Am I missing something?

```suggestion
  get pinned() {
    return this.data.pinned ?? null;
  }
```

File: packages/discord.js/src/structures/Message.js [hunk 2]
```diff
@@ -30,127 +30,401 @@ const Util = require('../util/Util');
...
-     * The timestamp the message was sent at
-     * @type {number}
+     * A random number or string used for checking message delivery
+     * <warn>This is only received after the message was sent successfully, and
+     * lost if re-fetched</warn>
+     * @type {?string}
      */
-    this.createdTimestamp = DiscordSnowflake.timestampFrom(this.id);
+    return this.data.nonce ?? null;
+  }
```
Comment (almeidx) [inline packages/discord.js/src/structures/Message.js:130]: The jsdoc should be above `get nonce()`

File: packages/discord.js/src/structures/Message.js [hunk 3]
```diff
@@ -30,127 +30,401 @@ const Util = require('../util/Util');
...
+
+  /**
+   * The message contents with all mentions replaced by the equivalent text.
+   * If mentions cannot be resolved to a name, the relevant mention in the message content will not be converted.
+   * @type {?string}
+   * @readonly
+   */
+  get cleanContent() {
+    // eslint-disable-next-line eqeqeq
+    return this.content != null ? Util.cleanContent(this.content, this.channel) : null;
```
Comment (almeidx) [inline packages/discord.js/src/structures/Message.js:305]: Why not do this instead? 

```suggestion
    return typeof this.content === 'string' ? Util.cleanContent(this.content, this.channel) : null;
```

File: packages/discord.js/src/structures/Message.js [hunk 4]
```diff
@@ -30,127 +30,401 @@ const Util = require('../util/Util');
...
+
+  /**
+   * Whether the message is editable by the client user
+   * @type {boolean}
+   * @readonly
+   */
+  get editable() {
+    const precheck = Boolean(this.author.id === this.client.user.id && (!this.guild || this.channel?.viewable));
+    // Regardless of permissions thread messages cannot be edited if
+    // the thread is locked.
```
Comment (almeidx) [inline packages/discord.js/src/structures/Message.js:316]:
```suggestion
    const precheck = Boolean(this.author.id === this.client.user.id && (!this.guild || this.channel?.viewable));
    // Regardless of permissions thread messages cannot be edited if the thread is locked.
```

**Review by almeidx — CHANGES_REQUESTED** [review]
I believe all the getters should have `@readonly` on their jsdocs

File: packages/discord.js/src/structures/Message.js [hunk 5]
```diff
@@ -30,127 +30,401 @@ const Util = require('../util/Util');
  * @extends {Base}
  */
 class Message extends Base {
-  constructor(client, data) {
+  constructor(client, data = {}) {
     super(client);
+    this.data.channelId = data.channel_id;
+    this.data.guildId = data.guild_id ?? this.channel?.guild?.id ?? null;
```
Comment (ckohen) [inline packages/discord.js/src/structures/Message.js:36]: data should stay snake_case probably, also you're not getting the camelCase versions later

**Review by ckohen — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/Message.js (hunk 5 above)
Comment (KhafraDev) [reply packages/discord.js/src/structures/Message.js:36]: Agreed, as message has guildId & channelId getters, there's no real purpose of storing camelCase properties.

Comment (suneettipirneni) [reply packages/discord.js/src/structures/Message.js:36]: Sorry, I thought I fixed this don't know why I marked this as resolved.

File: packages/discord.js/src/structures/Message.js [hunk 6]
```diff
@@ -439,8 +351,212 @@ class Message extends Base {
...
+      reactions,
+      ...initData
+    } = data;
+
+    /**
+     * The raw API data for this message
+     * @type {APIMessage}
+     * @readonly
+     */
+    this.data = { ...initData, ...this.data };
```
Comment (almeidx) [inline packages/discord.js/src/structures/Message.js:453]: If `APIMessage` is supposed to be the type from -types then it's not correct, as it's missing `attachments`, `components`, etc.

Comment (almeidx) [reply packages/discord.js/src/structures/Message.js:453]: Yeah, I meant `this.data` doesn't contain the things that I listed. So APIMessage is not the correct type for `this.data`

File: packages/discord.js/src/structures/Message.js [hunk 7]
```diff
@@ -373,16 +249,63 @@ class Message extends Base {
...
+      id: this.data.interaction.id,
+      type: this.data.interaction.type,
+      commandName: this.data.interaction.name,
+      user: this.client.users.resolve(this.data.interaction.user),
+    };
+  }
+
+  /**
+   * The channel that the message was sent in
+   * @type {TextChannel|DMChannel|NewsChannel|ThreadChannel}
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Message.js:295]:
```suggestion
   * @type {TextBasedChannels}
```

File: packages/discord.js/src/structures/Message.js [hunk 8]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-      /**
-       * Supplemental application information for group activities
-       * @type {?ClientApplication}
-       */
-      this.groupActivityApplication = new ClientApplication(this.client, data.application);
-    } else {
-      this.groupActivityApplication ??= null;
-    }
+  /**
+   * Whether or not this message is pinned
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Message.js:123]:
```suggestion
   * Whether this message is pinned
```

File: packages/discord.js/src/structures/Message.js [hunk 9]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-      /**
-       * Flags that are applied to the message
-       * @type {Readonly<MessageFlagsBitField>}
-       */
-      this.flags = new MessageFlagsBitField(data.flags).freeze();
-    } else {
-      this.flags = new MessageFlagsBitField(this.flags).freeze();
-    }
+  /**
+   * Whether or not the message was Text-To-Speech
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Message.js:164]:
```suggestion
   * Whether the message was Text-to-Speech
```

File: packages/discord.js/src/structures/Message.js (hunk 6 above)
Comment (suneettipirneni) [reply packages/discord.js/src/structures/Message.js:453]: Oh well I'm not really sure how to `Omit` in jsdoc (without listing every individual prop) so I left in the base type.

File: packages/discord.js/src/structures/Message.js [hunk 10]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-     * @typedef {Object} MessageReference
-     * @property {Snowflake} channelId The channel's id the message was referenced
-     * @property {?Snowflake} guildId The guild's id the message was referenced
-     * @property {?Snowflake} messageId The message's id that was referenced
-     */
+  /**
+   * A random number or string used for checking message delivery
+   * <warn>This is only received after the message was sent successfully, and
+   * lost if re-fetched</warn>
+   * @type {?string}
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Message.js:176]:
```suggestion
   * @type {?(string|number)}
```

File: packages/discord.js/src/structures/Message.js [hunk 11]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-      if (data.reactions?.length > 0) {
-        for (const reaction of data.reactions) {
-          this.reactions._add(reaction);
-        }
-      }
-    } else {
-      this.reactions ??= new ReactionManager(this);
-    }
+  /**
+   * Whether or not this message was sent by Discord, not actually a user (e.g. pin notifications)
```
Comment (almeidx) [inline packages/discord.js/src/structures/Message.js:95]:
```suggestion
   * Whether this message was sent by Discord, not actually a user (e.g. pin notifications)
```

Comment (imranbarbhuiya) [reply packages/discord.js/src/structures/Message.js]: System message

File: packages/discord.js/src/structures/Message.js [hunk 12]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-    } else {
-      this.reference ??= null;
-    }
+  /**
+   * The timestamp the message was last edited at (if applicable)
+   * @type {?number}
+   * @readonly
+   */
+  get editedTimestamp() {
+    if (this.data.edited_timestamp === undefined) return undefined;
```
Comment (imranbarbhuiya) [inline packages/discord.js/src/structures/Message.js:189]:
```suggestion
    if (this.data.edited_timestamp === undefined) return null;
```

File: packages/discord.js/src/structures/Message.js [hunk 13]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-
-    /**
-     * The id of the guild the message was sent in, if any
-     * @type {?Snowflake}
-     */
-    this.guildId = data.guild_id ?? this.channel?.guild?.id ?? null;
-
+    this.data.channel_id = data.channel_id;
+    this.data.guild_id = data.guild_id ?? this.channel?.guild?.id ?? null;
     this._patch(data);
```
Comment (imranbarbhuiya) [inline packages/discord.js/src/structures/Message.js:37]: `this.data` is undefined here. 
```suggestion
    this._patch(data);
    this.data.channel_id = data.channel_id;
    this.data.guild_id = data.guild_id ?? this.channel?.guild?.id ?? null;
```

will fix the issue ig

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/Message.js [hunk 14]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-      this.activity = {
-        partyId: data.activity.party_id,
-        type: data.activity.type,
-      };
-    } else {
-      this.activity ??= null;
-    }
+  /**
+   * A list of MessageActionRows in the message
+   * @type {ActionRow[]}
```
Comment (IRONM00N) [inline packages/discord.js/src/structures/Message.js:141]:
```suggestion
   * @type {ActionRow[]}
   * @readonly
```

File: packages/discord.js/src/structures/Message.js [hunk 15]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
-       * The id of the application of the interaction that sent this message, if any
-       * @type {?Snowflake}
-       */
-      this.applicationId = data.application_id;
-    } else {
-      this.applicationId ??= null;
-    }
+  /**
+   * A list of embeds in the message - e.g. YouTube Player
+   * @type {Embed[]}
```
Comment (IRONM00N) [inline packages/discord.js/src/structures/Message.js:133]:
```suggestion
   * @type {Embed[]}
   * @readonly
```

File: packages/discord.js/src/structures/Message.js [hunk 16]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
+  get components() {
+    return this._components ?? [];
+  }
 
-    if ('thread' in data) {
-      this.client.channels._add(data.thread, this.guild);
-    }
+  /**
+   * A collection of attachments in the message - e.g. Pictures - mapped by their ids
+   * @type {Collection<Snowflake, MessageAttachment>}
```
Comment (IRONM00N) [inline packages/discord.js/src/structures/Message.js:149]:
```suggestion
   * @type {Collection<Snowflake, MessageAttachment>}
   * @readonly
```

File: packages/discord.js/src/structures/Message.js [hunk 17]
```diff
@@ -30,340 +30,216 @@ const Util = require('../util/Util');
...
+  }
 
-    if (this.member && data.member) {
-      this.member._patch(data.member);
-    } else if (data.member && this.guild && this.author) {
-      this.guild.members._add(Object.assign(data.member, { user: this.author }));
-    }
+  /**
+   * A collection of stickers in the message
+   * @type {Collection<Snowflake, Sticker>}
```
Comment (IRONM00N) [inline packages/discord.js/src/structures/Message.js:157]:
```suggestion
   * @type {Collection<Snowflake, Sticker>}
   * @readonly
```

File: packages/discord.js/typings/index.d.ts [hunk 18]
```diff
@@ -1498,46 +1498,60 @@ export class Message<Cached extends boolean = boolean> extends Base {
...
-  public type: MessageType;
+  public readonly tts: boolean;
+  public readonly type: MessageType;
   public readonly url: string;
-  public webhookId: Snowflake | null;
-  public flags: Readonly<MessageFlagsBitField>;
-  public reference: MessageReference | null;
+  public readonly webhookId: Snowflake | null;
+  public readonly flags: Readonly<MessageFlagsBitField>;
+  public readonly reference: MessageReference | null;
```
Comment (IRONM00N) [inline packages/discord.js/typings/index.d.ts:1554]: Use use getter notation

**Review by IRONM00N — CHANGES_REQUESTED** [review]

**Pull request closed without merging**